File: app/src/main/java/com/rafalwesolowski/testapp/realm/DefaultRealmManager.java
package com.rafalwesolowski.testapp.realm;

import java.util.List;

import javax.inject.Inject;

import io.realm.Case;
import io.realm.Realm;
import io.realm.exceptions.RealmException;

/**
 * Created by rafalwesolowski on 21/01/2018.
 */

public class DefaultRealmManager implements RealmManager {

    private final Realm mRealm;

    @Inject
    public DefaultRealmManager(Realm realm) {
        mRealm = realm;
    }

    @Override
    public void saveTask(RealmTask task) {
        try {
            if (task != null) {
                mRealm.beginTransaction();
                RealmTask realmTask = getTask(task.getTitle());
                if (realmTask == null) {
                    realmTask = mRealm.createObject(RealmTask.class);
                }

                if (realmTask != null) {
                    realmTask.setTitle(task.getTitle());
                }

                mRealm.commitTransaction();
            }
        } catch (RealmException e) {
            mRealm.cancelTransaction();
        }
    }

    @Override
    public void deleteTask(RealmTask task) {
        try {
            if (task != null) {
                mRealm.beginTransaction();

                mRealm.where(RealmTask.class)
                        .equalTo(RealmTask.FIELD_NAME, task.getTitle(), Case.SENSITIVE)
                        .findAll()
                        .clear();

                mRealm.commitTransaction();
            }
        } catch (RealmException e) {
            mRealm.cancelTransaction();
        }
    }

    @Override
    public void saveTaskWithTitle(String taskTitle) {
        RealmTask realmTask = new RealmTask(taskTitle);
        saveTask(realmTask);
    }

    @Override
    public List<RealmTask> getAllTasks() {
        return mRealm.where(RealmTask.class).findAllSorted(RealmTask.FIELD_NAME);
    }

    private RealmTask getTask(String taskName) {
        return mRealm.where(RealmTask.class)
                .equalTo(RealmTask.FIELD_NAME, taskName, Case.SENSITIVE)
                .findFirst();
    }
}

File: app/src/main/java/com/rafalwesolowski/testapp/realm/DefaultRealmManager.kt
package com.rafalwesolowski.testapp.realm

import javax.inject.Inject

import io.realm.Case
import io.realm.Realm
import io.realm.exceptions.RealmException

/**
 * Created by rafalwesolowski on 21/01/2018.
 */

class DefaultRealmManager @Inject
constructor(private val mRealm: Realm) : RealmManager {

    override fun saveTask(task: RealmTask) {
        try {
            mRealm.beginTransaction()
            var realmTask = getTask(task.title)
            if (realmTask == null) {
                realmTask = mRealm.createObject(RealmTask::class.java)
            }

            realmTask?.title = task.title

            mRealm.commitTransaction()
        } catch (e: RealmException) {
            mRealm.cancelTransaction()
        }

    }

    override fun deleteTask(task: RealmTask) {
        try {
            mRealm.beginTransaction()

            mRealm.where(RealmTask::class.java)
                    .equalTo(RealmTask.FIELD_NAME, task.title, Case.SENSITIVE)
                    .findAll()
                    .clear()

            mRealm.commitTransaction()
        } catch (e: RealmException) {
            mRealm.cancelTransaction()
        }

    }

    override fun saveTaskWithTitle(taskTitle: String) {
        val realmTask = RealmTask()
        realmTask.title = taskTitle
        saveTask(realmTask)
    }

    override fun getAllTasks(): List<RealmTask> {
        return mRealm.where(RealmTask::class.java).findAllSorted(RealmTask.FIELD_NAME)
    }

    private fun getTask(taskName: String?): RealmTask? {
        return mRealm.where(RealmTask::class.java)
                .equalTo(RealmTask.FIELD_NAME, taskName, Case.SENSITIVE)
                .findFirst()
    }
}
